packages/sources/coinpaprika-state/src/transport/coinpaprika-state.ts

```diff
@@ -0,0 +1,238 @@
+ */
+export class CoinpaprikaStateTransport extends StreamingTransport<TransportTypes> {
+  name!: string
+  responseCache!: ResponseCache<TransportTypes>
+  requester!: Requester
+
+  private connectionManager!: SSEConnectionManager
+  private activePairs: Map<string, RequestParams> = new Map()
+
+  async initialize(
+    dependencies: TransportDependencies<TransportTypes>,
+    adapterSettings: TransportTypes['Settings'],
+    endpointName: string,
+    transportName: string,
+  ): Promise<void> {
+    await super.initialize(dependencies, adapterSettings, endpointName, transportName)
+    this.requester = dependencies.requester
+    this.connectionManager = new SSEConnectionManager(this.requester, COINPAPRIKA_STATE_EVENT_TYPE)
+  }
+
+  async streamHandler(
+    context: EndpointContext<TransportTypes>,
+    subscriptions: SubscriptionDeltas<TypeFromDefinition<TransportTypes['Parameters']>>,
+  ): Promise<void> {
+    if (
+      subscriptions.new.length ||
+      subscriptions.stale.length ||
+      !this.connectionManager.connected
+    ) {
+      const currentPairs = new Map<string, RequestParams>()
+      for (const sub of subscriptions.desired) {
+        currentPairs.set(`${sub.base}/${sub.quote}`, sub)
+      }
+
+      logger.debug(`Updating stream: ${currentPairs.size} pairs`)
+      await this.updateConnection(context, currentPairs)
+    }
+    await sleep(context.adapterSettings.BACKGROUND_EXECUTE_MS)
+  }
+
+  private async updateConnection(
+    context: EndpointContext<TransportTypes>,
+    pairs: Map<string, RequestParams>,
+  ): Promise<void> {
+    logger.debug(`updateConnection called with ${pairs.size} pairs`)
+
+    // update active pairs
+    this.activePairs = new Map(pairs)
+
+    if (pairs.size === 0) {
+      logger.debug('No pairs, disconnecting')
+      await this.connectionManager.disconnect()
+      return
+    }
+
+    const pairRequests: PairRequest[] = Array.from(pairs.values()).map((p) => ({
+      base: p.base,
+      quote: p.quote,
+    }))
+
+    const config: SSEConnectionConfig = {
+      apiEndpoint: context.adapterSettings.API_ENDPOINT,
+      apiKey: context.adapterSettings.API_KEY,
+      requestTimeoutMs: context.adapterSettings.REQUEST_TIMEOUT_MS,
+    }
+
+    const callbacks: SSEConnectionCallbacks = {
+      onData: async (eventType: string, data: string) => {
+        await this.handleSSEEvent(eventType, data)
+      },
+      onError: (error: Error) => {
+        logger.error(`SSE stream error: ${error.message}`)
+      },
+      onConnectionError: async (status: number) => {
+        await this.handleConnectionError(status)
+      },
+      onReconnectNeeded: () => {
+        logger.info('SSE ended; will reconnect on next cycle')
+      },
+    }
+
+    await this.connectionManager.connect(pairRequests, config, callbacks)
+  }
+
+  private async handleConnectionError(status: number): Promise<void> {
+    const mappedStatus = status === 500 ? 502 : status
```

Comment:
This should be the status from the adapter and not the provider, so if the provider is sending an errored response most often that would mean a 502 directly unless we can actually determine directly what the problem is and forward accordingly (for example, if from the provider response we can tell that outbound request was improperly formatted due to a bad paramter in the inbound EA request, we can put a 400 here)